### DISCUSSION TITLE
Package eleventy into a single binary or executable

### DISCUSSION THREAD

**Discussion by chrisdmacrae** [11ty/eleventy#3969] [Enhancement Queue]
One of the common complaints against almost all other static site generators is installation.

This is especially a problem for Jekyll on Windows.

Zeit/pkg creates application executables of node applications for Darwin, Win32, and Unix systems, and will still allow users to add plugins from the NPM ecosystem.

This should be a late roadmap goal, but I believe it will separate Eleventy from the pack by making it accessible to non-technical users.

This could be deployed to OS-level package managers (e.g, homebrew, package-manager, and apt) in order to make installation and update management a breeze, while still leaving the product as a CLI.

**Comment by chrisdmacrae** [comment]
Note, they are performance considerations here. Packaged node apps may run 1.5x slower than unpackaged apps.

This may not be a problem for eleventy as the major bottleneck is file I/O.

---

Correction, the v8 runtime in node v8.x.x+ now does compile-time optimizations that should make performance negligible from unpackaged apps.

**Comment by zachleat** [comment]
cc @mattdsteele this might alleviate the problem you were going through with requiring Node 6 support

**Comment by zachleat** [comment]
I know Hugo provides binaries and they have been successful there https://github.com/gohugoio/hugo/releases

I like this idea!

**Comment by zachleat** [comment]
https://github.com/nexe/nexe

**Comment by zachleat** [comment]
This repository is now using lodash style issue management for enhancements (see https://twitter.com/samselikoff/status/991395669016436736)

This means enhancement issues will now be closed instead of leaving them open. The enhancement backlog can be found here: https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+

**Comment by aral** [comment]
Just wanted to add that having a single binary would also make it much easier to include 11ty support in other tools like, *wink, wink* [Site.js](https://sitejs.org) alongside other single-binary static site generators like Hugo and Zola.

**Reply by aral** [reply]
A quick update since the issue is a discussion now, in case it helps: the single binary approach became an encumbrance with Site.js and slowed down deployments.

Instead, in Kitten (https://kitten.small-web.org), I have an installation script that downloads the exact Node + npm binary it needs. This is way more flexible and packaging/deployment are way faster and it still means the person doesn’t have to have a working Node install on their system to run it or worry about Node versioning (the real wins).

Please feel free to yoink the approach if you like :)

**Comment by zachleat** [comment]
Adding a link to https://nodejs.org/api/single-executable-applications.html via duplicate #4111
